discovery/sync_manager.go

```diff
@@ -0,0 +1,673 @@
+	DefaultActiveSyncerTimeout = 5 * time.Minute
+)
+
+var (
+	// ErrSyncManagerExiting is an error returned when we attempt to
+	// start/stop a gossip syncer for a connected/disconnected peer, but the
+	// SyncManager has already been stopped.
+	ErrSyncManagerExiting = errors.New("sync manager exiting")
+)
+
+// SyncManagerCfg contains all of the dependencies required for the SyncManager
+// to carry out its duties.
+type SyncManagerCfg struct {
+	// ChainHash is a hash that indicates the specific network of the active
+	// chain.
+	ChainHash chainhash.Hash
+
+	// ChanSeries is an interface that provides access to a time series view
+	// of the current known channel graph. Each GossipSyncer enabled peer
+	// will utilize this in order to create and respond to channel graph
+	// time series queries.
+	ChanSeries ChannelGraphTimeSeries
+
+	// NumActiveSyncers is the number of peers for which we should have
+	// active syncers with. After reaching NumActiveSyncers, any future
+	// gossip syncers will be passive.
+	NumActiveSyncers int
+
+	// RotateTicker is a ticker responsible for notifying the SyncManager
+	// when it should rotate its active syncers. A single active syncer with
+	// a chansSynced state will be exchanged for a passive syncer in order
+	// to ensure we don't keep syncing with the same peers.
+	RotateTicker ticker.Ticker
+
+	// HistoricalSyncTicker is a ticker responsible for notifying the
+	// SyncManager when it should attempt a historical sync with a gossip
+	// sync peer.
+	HistoricalSyncTicker ticker.Ticker
+
+	// ActiveSyncerTimeoutTicker is a ticker responsible for notifying the
+	// SyncManager when it should attempt to start the next pending
+	// activeSyncer due to the current one not completing its state machine
+	// within the timeout.
+	ActiveSyncerTimeoutTicker ticker.Ticker
+}
+
+// SyncManager is a subsystem of the gossiper that manages the gossip syncers
+// for peers currently connected. When a new peer is connected, the manager will
+// create its accompanying gossip syncer and determine whether it should have an
+// ActiveSync or PassiveSync sync type based on how many other gossip syncers
+// are currently active. Any ActiveSync gossip syncers are started in a
+// round-robin manner to ensure we're not syncing with multiple peers at the
+// same time. The first GossipSyncer registered with the SyncManager will
+// attempt a historical sync to ensure we have as much of the public channel
+// graph as possible.
+type SyncManager struct {
+	start sync.Once
+	stop  sync.Once
+
+	cfg SyncManagerCfg
+
+	// historicalSync allows us to perform an initial historical sync only
+	// _once_ with a peer during the SyncManager's startup.
+	historicalSync sync.Once
+
+	// activeSyncers is the set of all syncers for which we are currently
+	// receiving graph updates from. The number of possible active syncers
+	// is bounded by NumActiveSyncers.
+	activeSyncers map[routing.Vertex]*GossipSyncer
```

Comment:
The mutex is still needed between the `syncHandler` and `roundRobinHandler` in order to determine whether it should be queued or not.

> The pendingActiveSyncers map is not needed, as you can remove directly from the list.

This was suggested [here](https://github.com/lightningnetwork/lnd/pull/2740#discussion_r268017032) to avoid an O(N) lookup and slice tricks when removing a stale active syncer, but agree that it's not really needed.

Reply:
Resolved offline.